### PR TITLE
Update MSC3786 implementation: Check the `state_key`

### PR DESCRIPTION
**Pull request by SimonBrandner**
MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/3786

<hr>

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: changelog.d/12939.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+MSC3786: Check `state_key`.
```
Comment (squahtx) [inline changelog.d/12939.feature:1]: I'd throw in some extra words for clarity:
```suggestion
Update MSC3786 implementation to check `state_key`.
```

**Review by squahtx — COMMENTED** [review]
lgtm, one small suggestion

File: changelog.d/12939.feature (hunk 1 above)
Comment (clokep) [reply changelog.d/12939.feature:1]: Also -- I think this is a bugfix, not a feature?

**Review by squahtx — APPROVED** [review]
LGTM!

**Pull request merged**